File: src/main/java/tree/TreeChild.java
package tree;


import java.util.ArrayList;
import java.util.List;

/**
 * 子节点链表示法：每个父节点记住它所有的子节点，在本例中标记为子节点链
 * 使用子节点链表示法来表示一棵树
 */
public class TreeChild<E> {

    // 节点的子链元素
    private static class SonNode {
        // 记录当前节点的位置
        private int pos;
        private SonNode next;

        public  SonNode(int pos, SonNode next) {
            this.pos = pos;
            this.next = next;
        }
    }

    // 树的节点元素
    public static class Node<T> {
        T data;
        // 记录第一个子节点
        SonNode first;

        public Node(T data) {
            this.data = data;
            this.first = null;
        }

        @Override
        public String toString() {

            if (first != null) {
                return "TreeChild$Node{" +
                        "data=" + data +
                        ", first=" + first.pos +
                        '}';
            } else {
                return "TreeChild$Node{" +
                        "data=" + data +
                        ", first= -1" +
                        '}';
            }
        }
    }

    private final int DEFAULT_TREE_SIZE = 100;
    private int treeSize = 0;
    private Node<E>[] nodes;
    // 记录节点数
    private int nodeNum;

    // 以指定根结点创建树
    public TreeChild(E data) {
        treeSize = DEFAULT_TREE_SIZE;
        nodes = new Node[treeSize];
        nodes[0] = new Node<>(data);
        nodeNum++;
    }

    public TreeChild(E data, int treeSize) {
        this.treeSize = treeSize;
        nodes = new Node[treeSize];
        nodes[0] = new Node<>(data);
        nodeNum++;
    }

    // 为指定节点添加子节点
    public void addNode(E data, Node parent) {
        for (int i = 0; i < treeSize; i++) {
            // 找到数组中第一个为 null 的元素，该元素保存新节点
            if (nodes[i] == null) {
                // 创建新的节点，并用指定数组元素来保存它
                nodes[i] = new Node<>(data);
                // 子节点链为空时
                if (parent.first == null) {
                    parent.first = new SonNode(i, null);
                } else {
                    SonNode next = parent.first;
                    // 子节点链不为空时，循环至子节点链的最后一个节点
                    while (next.next != null) {
                        next = next.next;
                    }
                    next.next = new SonNode(i, null);
                }
                nodeNum++;
                return;
            }
        }
        throw new RuntimeException("该树已满，无法添加新的节点");
    }

    public boolean empty() {
        return nodes[0] == null;
    }

    public Node<E> root() {
        return nodes[0];
    }

    // 获得指定节点的所有子节点
    public List<Node<E>> children(Node parent) {
        List<Node<E>> list = new ArrayList<>();
        SonNode next = parent.first;
        while (next != null) {
            list.add(nodes[next.pos]);
            next = next.next;
        }
        return list;
    }

    // 获得指定子节点的第 index 个子节点
    public Node<E> child(Node parent, int index) {
        SonNode next = parent.first;
        for (int i = 0; next != null; i++) {
            if (index == i) {
                return nodes[next.pos];
            }
            next = next.next;
        }
        return null;
    }

    // 获得树的深度
    public int deep() {
        return deep(root());
    }

    // 递归方法：每棵树的深度为其子树的最大深度 + 1
    private int deep(Node node) {
        if (node.first == null) {
            return 1;
        } else {
            int max = 0;
            SonNode next = node.first;
            while (next != null) {
                int tmp = deep(nodes[next.pos]);
                if (tmp > max) {
                    max = tmp;
                }
                next = next.next;
            }
            return max + 1;
        }
    }


    public int pos(Node node) {
        for (int i = 0; i < treeSize; i++) {
            if (nodes[i] == node) {
                return i;
            }
        }
        return -1;
    }
}

File: src/main/java/tree/TreeChild.kt
package tree

class TreeChild<E>(data: E) {
    private val DEFAULT_TREE_SIZE = 100
    private var treeSize = 0
    private var nodes: Array<Node<E>?>
    private var nodeNums = 0

    init {
        treeSize = DEFAULT_TREE_SIZE
        nodes = arrayOfNulls(this.treeSize)
        nodes[0] = Node(data)
        nodeNums++
    }

    constructor(data: E, treeSize: Int) : this(data) {
        this.treeSize = treeSize
    }

    fun addNode(data: E, parent: Node<E>) {
        for ((index, value) in nodes.withIndex()) {
            // 找对数组中第一个为 null 的元素，该元素保存新的节点
            if (value == null) {
                // 将数组保存到数组中，在接下来的操作中，建立节点之间的关系
                nodes[index] = Node(data)
                if (parent.first == null) {
                    parent.first = SonNode(index, null)
                } else {
                    // 遍历子链
                    var next = parent.first
                    while (next?.next != null) {
                        next = next.next
                    }
                    next?.next = SonNode(index, null)
                }
                nodeNums++
                return
            }
        }
        throw RuntimeException("该树已满，无法添加新的节点")
    }

    fun empty(): Boolean = nodes[0] == null

    fun root(): Node<E> = nodes[0]!!

    /**
     * 返回指定节点（非叶子节点）的所有子节点
     */
    fun children(parent: Node<E>): List<Node<E>> {
        val list = mutableListOf<Node<E>>()
        var next = parent.first
        while (next != null) {
            list.add(nodes[next.pos]!!)
            next = next.next
        }
        return list
    }

    /**
     * 返回指定节点的第 index 个子节点
     */
    fun child(parent: Node<E>, index: Int): Node<E>? {
        var next = parent.first
        var i:Int = 0
        while (next != null){
            if (i == index) {
                return nodes[next.pos]
            }
            next = next.next
            i++
        }
        return null
    }

    fun deep(): Int = deep(root())

    /**
     * 获得数的深度
     */
    private fun deep(node: Node<E>): Int {
        if (node.first == null) {
            return 1
        } else {
            var max = 0
            var next = node.first
            while (next != null) {
                var tmp = deep(nodes[next.pos]!!)
                if (tmp > max) {
                    max = tmp
                }
                next = next.next

            }
            return max + 1
        }
    }

    fun pos(node: Node<E>): Int {
        for (index in nodes.indices) {
            if (nodes[index] == node) {
                return index
            }
        }
        return -1
    }

    companion object {
        // 节点 Node中如果拥有子链，那么 SonNode 通过 next 用来维护 子链的关系，以及子链元素在数组中的索引
        class SonNode(val pos: Int, var next: SonNode?)
        // 树中的节点，
        class Node<T>(private val data: T) {
            // 第一个子节点
            var first: SonNode? = null
            override fun toString(): String = if (first != null) {
                "SonNode(data=$data, first.pos=${first?.pos})"
            } else {
                "SonNode(data=$data, first.pos=-1)"
            }
        }

    }
}

fun main(args: Array<String>) {
    val tree = TreeChild("", 2)
}